File: cpp/hamming/hamming.cpp
#include "hamming.h"
#include <string>

using namespace std;

namespace hamming {
    unsigned int compute(string a, string b) {
        if (a.size() != b.size()) {
            throw domain_error("Strand lengths must be equal!");
        }

        unsigned int distance = 0;
        for (unsigned int i = 0; i < a.size(); i++) {
            distance += a[i] != b[i];
        }

        return distance;
    }
}  // namespace hamming

File: go/hamming/hamming.go
package hamming

import "errors"

func Distance(a, b string) (int, error) {
	// Transform first to an array of runes to deal with multi-byte characters
	ra := []rune(a)
	rb := []rune(b)

	if len(ra) != len(rb) {
		return -1, errors.New("Both strings must have equal lengths")
	}

 	distance := 0
	for i := range ra {
		if ra[i] != rb[i] {
			distance += 1
		}
	}
	return distance, nil
}
